Fix IntegrationFlowContext concurrency issue

artembilan opened the pull request:
When we register `IntegrationFlow` s concurrently at runtime, we may
end up with the problem when we register the same object with the same
bean name, but in different places.
Or when we turn off bean overriding, we end up with the exception that
bean with the name already registered

* Wrap `IntegrationFlow` bean registration in the
`StandardIntegrationFlowContext` into the `Lock` when its bean name
is generating
* Make `StandardIntegrationFlowContext.registry` as `ConcurrentHashMap`
to avoid `ConcurrentModificationException` during `put()` and `remove()`
* Fix concurrency for beans registration with the generation names in
the `IntegrationFlowBeanPostProcessor` using an `IntegrationFlow` id
as a prefix for uniqueness.

**Cherry-pick to 5.0.x**

artembilan committed:
Fix IntegrationFlowContext concurrency issue

When we register `IntegrationFlow` s concurrently at runtime, we may
end up with the problem when we register the same object with the same
bean name, but in different places.
Or when we turn off bean overriding, we end up with the exception that
bean with the name already registered

* Wrap `IntegrationFlow` bean registration in the
`StandardIntegrationFlowContext` into the `Lock` when its bean name
is generating
* Make `StandardIntegrationFlowContext.registry` as `ConcurrentHashMap`
to avoid `ConcurrentModificationException` during `put()` and `remove()`
* Fix concurrency for beans registration with the generation names in
the `IntegrationFlowBeanPostProcessor` using an `IntegrationFlow` id
as a prefix for uniqueness.

**Cherry-pick to 5.0.x**

artembilan commented:
See https://stackoverflow.com/questions/50393906/spring-integration-java-dsl-dynamically-create-integrationflows.

And also https://jira.spring.io/browse/SPR-16837

artembilan commented:
I think there is no reason in locking/synchronization in the `IntegrationFlowBeanPostProcessor` - we simply can generate bean names based on the flow id as a prefix.

Will come back with such a fix later today...

artembilan commented:
Now it's ready for review and merging.

Thanks

garyrussell commented:
Do you think the travis failure is legitimate, or should I try to build locally?

artembilan committed:
Fix generated bean name in the WebFluxDslTests

garyrussell commented:
I am a bit uncomfortable cherry-picking the bean name change to 5.0.x since it's a breaking change. But perhaps we have no choice.

We should add something to the what's new, DSL chapter, and migration guide(s) for the bean name change.

artembilan commented:
Right, will do.

Thanks for the reminders!

artembilan committed:
Use only single `Lock` in the `StandardIntegrationFlowContext`:
we don't need a fully blown `LockRegistry` there anymore since we have
only one synchronization block there and it is always around the same
type
* Add `What's New` note, and mention changes in the `dsl.adoc`

artembilan commented:
I will add Migration Guide note after this is merged.

garyrussell commented:
Doc polishing for your review: https://github.com/garyrussell/spring-integration/commit/e572c682efe2d709c455d13806d3a951bc85f852

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/b21dc0dff353d735f7acf915ab189550297c2ceb.

Cherry-picked to `5.0.x` as: https://github.com/spring-projects/spring-integration/commit/d0603a80ff5d89d14681ed6acec17c94c08e9564
